firedrake/progress_bar.py

```diff
@@ -0,0 +1,76 @@
+from pyop2.mpi import COMM_WORLD
+from progress.bar import FillingSquaresBar
+
+
+class _NullProgressBar:
```

Comment:
Is this class needed? It's identical to the one in [pyadjoint](https://github.com/dolfin-adjoint/pyadjoint/blob/master/pyadjoint/tape.py#L467).

Reply:
I guess we can only use the pyadjoint equivalent if pyadjoint is a hard dependency, otherwise we need to have the same class in both places.

Reply:
pyadjoint is a hard dependency but I was weighing up using a private class from outside the package versus repeating a small amount of code. I came down on the side of repeating the code, but if people feel strongly the other way then we can import the class. I don't really care very much either way.

Reply:
I personally don't have a strong opinion either and can think of plusses and minuses to both (e.g. where one to find a bug). @connorjward ?

Reply:
I have no strong opinion either. Happy to leave as is.